File: Windows/App/OAuth.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace HelmsleyDrive.App;

/// <summary>
/// Talks to the portal's OAuth endpoints (<c>backend/routes/oauth/</c>), mounted at the site
/// root, so <c>/authorize</c> and <c>/token</c> are absolute paths rather than anything under
/// <c>/api</c>.
/// </summary>
public static class OAuth
{
    sealed class TokenResponse
    {
        [JsonPropertyName("access_token")] public string? AccessToken { get; set; }
        [JsonPropertyName("refresh_token")] public string? RefreshToken { get; set; }
        [JsonPropertyName("expires_in")] public double? ExpiresIn { get; set; }
    }

    /// <summary>
    /// A code verifier and its S256 challenge. The whole of a public client's security: the app
    /// ships with no secret it could keep, so what proves the code is being redeemed by the process
    /// that requested it is knowledge of a value only that process generated.
    /// </summary>
    public sealed class Pkce
    {
        public string Verifier { get; } = Base64Url(RandomNumberGenerator.GetBytes(32));
        public string Challenge => Base64Url(SHA256.HashData(Encoding.ASCII.GetBytes(Verifier)));
    }

    public static string AuthorizeUrl(Pkce pkce, string state)
    {
        var query = new Dictionary<string, string>
        {
            ["response_type"] = "code",
            ["client_id"] = Configuration.OAuthClientId,
            ["redirect_uri"] = Configuration.OAuthRedirectUri,
            ["code_challenge"] = pkce.Challenge,
            ["code_challenge_method"] = "S256",
            ["state"] = state,
            ["scope"] = Configuration.OAuthScope,
        };
        var encoded = string.Join("&", query.Select(p => $"{p.Key}={Uri.EscapeDataString(p.Value)}"));
        return new Uri(Configuration.BaseUri, "authorize").ToString() + "?" + encoded;
    }

    public static Task<TokenSet> Exchange(string code, string verifier) =>
        Token(new Dictionary<string, string>
        {
            ["grant_type"] = "authorization_code",
            ["code"] = code,
            ["redirect_uri"] = Configuration.OAuthRedirectUri,
            ["client_id"] = Configuration.OAuthClientId,
            ["code_verifier"] = verifier,
        });

    /// <summary>
    /// Whether a token-endpoint refusal says the grant itself is dead.
    ///
    /// <c>invalid_grant</c> is the only refusal that means the refresh token is spent and signing
    /// in again is the way back. Everything else a 400 can carry — <c>invalid_request</c>, a
    /// proxy's error page, a body that is not JSON at all — leaves a refresh token that is still
    /// good, so an unrecognised refusal reads as "not this", and the credential is kept.
    /// </summary>
    public static bool IsInvalidGrant(string body)
    {
        try
        {
            using var parsed = JsonDocument.Parse(body);
            return parsed.RootElement.TryGetProperty("error", out var error) && error.GetString() == "invalid_grant";
        }
        catch (JsonException)
        {
            return false;
        }
    }

    public static Task<TokenSet> Refresh(string refreshToken) =>
        Token(new Dictionary<string, string>
        {
            ["grant_type"] = "refresh_token",
            ["refresh_token"] = refreshToken,
            ["client_id"] = Configuration.OAuthClientId,
        });

    // The token endpoint takes no bearer token, so it wants none of the API client's plumbing —
    // but redirects are off here too, and for a sharper reason than over there. A 307 or 308
    // replays the POST body verbatim, and this body carries the rotating refresh token: the
    // credential that outlives every access token this app will ever hold. The endpoint answering
    // with a redirect at all is a fault worth surfacing rather than following.
    static readonly HttpClient Http = new(new SocketsHttpHandler
    {
        AllowAutoRedirect = false,
        PooledConnectionLifetime = TimeSpan.FromMinutes(5),
    });

    static async Task<TokenSet> Token(Dictionary<string, string> form)
    {
        using var response = await Http.PostAsync(
            new Uri(Configuration.BaseUri, "token"),
            new FormUrlEncodedContent(form)).ConfigureAwait(false);
        var body = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
        if (!response.IsSuccessStatusCode)
            throw new OAuthException((int)response.StatusCode, body);

        var decoded = JsonSerializer.Deserialize<TokenResponse>(body);
        if (decoded?.AccessToken is not { } access)
            throw new OAuthException((int)response.StatusCode, "The sign-in server sent a response this app could not read.");
        // The refresh token rotates on every use server-side, so a response without one would leave
        // nothing to refresh with next time — better to fail here than to discover it in an hour.
        if (decoded.RefreshToken is not { } refresh)
            throw new OAuthException((int)response.StatusCode, "The sign-in server issued no refresh token.");
        return new TokenSet(access, refresh, DateTimeOffset.UtcNow.AddSeconds(decoded.ExpiresIn ?? 3600));
    }

    /// <summary>RFC 7636 wants base64url with no padding, which Convert's base64 is not.</summary>
    public static string Base64Url(byte[] bytes) =>
        Convert.ToBase64String(bytes).Replace('+', '-').Replace('/', '_').TrimEnd('=');
}

/// <summary>The sign-in server refused the request.</summary>
public sealed class OAuthException(int status, string body)
    : Exception($"The sign-in server refused the request (HTTP {status}): {body}")
{
    public int Status { get; } = status;
    public string Body { get; } = body;
}

public sealed class NotAuthenticatedException()
    : Exception("Not signed in to Helmsley. Run Helmsley Drive and sign in again.");

/// <summary>
/// Hands out an access token that is valid right now, refreshing when it is not.
///
/// Serialised, because hydrations arrive concurrently and a refresh must happen once: the server
/// rotates refresh tokens, so two simultaneous refreshes would see the second revoke what the
/// first had just been issued. The Mac side also re-reads the store mid-refresh in case the other
/// process rotated the pair; there is no other process here, so that recovery has no counterpart.
/// </summary>
public sealed class TokenProvider
{
    public static readonly TokenProvider Shared = new();

    readonly SemaphoreSlim _refreshing = new(1, 1);

    // Guards _cached and _generation only. Distinct from the refresh gate, which is held across a
    // network round trip and so can never be taken by a sign-out on the way past.
    readonly object _gate = new();

    TokenSet? _cached;

    public bool IsSignedIn => (_cached ?? TokenStore.Load()) is not null;

    // Bumped by every sign-in and sign-out. A refresh in flight captures it and checks it again
    // before caching what it was given: a sign-out that lands mid-refresh would otherwise have the
    // refresh's own `_cached = refreshed` land after it and write the credential back to disk,
    // leaving a signed-out process holding a live token.
    int _generation;

    public void Store(TokenSet tokens)
    {
        lock (_gate) { _generation++; _cached = tokens; }
        TokenStore.Save(tokens);
    }

    public void SignOut()
    {
        lock (_gate) { _generation++; _cached = null; }
        TokenStore.Clear();
    }

    public async Task<string> AccessToken()
    {
        await _refreshing.WaitAsync().ConfigureAwait(false);
        try
        {
            int generation;
            lock (_gate)
            {
                if (_cached is { IsAccessTokenFresh: true } cached) return cached.AccessToken;
                generation = _generation;
            }

            var stored = TokenStore.Load() ?? throw new NotAuthenticatedException();
            if (stored.IsAccessTokenFresh)
            {
                lock (_gate) { if (_generation == generation) _cached = stored; }
                return stored.AccessToken;
            }

            try
            {
                Console.WriteLine("refreshing access token");
                var refreshed = await OAuth.Refresh(stored.RefreshToken).ConfigureAwait(false);
                lock (_gate)
                {
                    // Somebody signed out while this was in flight. The token is good and is
                    // deliberately thrown away: writing it back would undo their sign-out.
                    if (_generation != generation) throw new NotAuthenticatedException();
                    // Cached before the save: the old refresh token is already spent, so if the
                    // save fails this memory is the only copy of the credential there is.
                    _cached = refreshed;
                }
                try { TokenStore.Save(refreshed); }
                catch (Exception e) { Console.Error.WriteLine($"token save failed (signed in until exit): {e.Message}"); }
                return refreshed.AccessToken;
            }
            catch (OAuthException e) when (e.Status is 400 or 401 && OAuth.IsInvalidGrant(e.Body))
            {
                // The server says the grant itself is dead — not that the request was malformed or
                // that a proxy hiccuped, which also arrive as 400s and must not cost the
                // credential. There is nothing left but to sign in again.
                SignOut();
                throw new NotAuthenticatedException();
            }
        }
        finally
        {
            _refreshing.Release();
        }
    }
}

File: Mac/Shared/OAuth.swift
import CryptoKit
import Foundation

/// Talks to the portal's OAuth endpoints (`backend/routes/oauth/`), mounted at the site root, so
/// `/authorize` and `/token` are absolute paths rather than anything under `/api`.
enum OAuth {

    struct TokenResponse: Decodable {
        let access_token: String
        let refresh_token: String?
        let expires_in: Double?
    }

    /// A code verifier and its S256 challenge. The whole of a public client's security: the app
    /// ships to laptops with no secret it could keep, so what proves the code is being redeemed by
    /// the process that requested it is knowledge of a value only that process generated.
    struct PKCE {
        let verifier: String
        var challenge: String {
            let digest = SHA256.hash(data: Data(verifier.utf8))
            return Data(digest).base64URLEncodedString()
        }

        init() {
            var bytes = [UInt8](repeating: 0, count: 32)
            _ = SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes)
            verifier = Data(bytes).base64URLEncodedString()
        }
    }

    static func authorizeURL(pkce: PKCE, state: String) -> URL {
        var components = URLComponents(
            url: Configuration.baseURL.appendingPathComponent("authorize"),
            resolvingAgainstBaseURL: false
        )!
        components.queryItems = [
            URLQueryItem(name: "response_type", value: "code"),
            URLQueryItem(name: "client_id", value: Configuration.oauthClientID),
            URLQueryItem(name: "redirect_uri", value: Configuration.oauthRedirectURI),
            URLQueryItem(name: "code_challenge", value: pkce.challenge),
            URLQueryItem(name: "code_challenge_method", value: "S256"),
            URLQueryItem(name: "state", value: state),
            // What this app is for: 'drive' opens the file-sync API and nothing else. The portal grants
            // by our registration either way, but asking for what we mean keeps the consent honest.
            URLQueryItem(name: "scope", value: "drive"),
        ]
        return components.url!
    }

    static func exchange(code: String, verifier: String) async throws -> TokenSet {
        try await token(form: [
            "grant_type": "authorization_code",
            "code": code,
            "redirect_uri": Configuration.oauthRedirectURI,
            "client_id": Configuration.oauthClientID,
            "code_verifier": verifier,
        ])
    }

    /// Whether a token-endpoint refusal says the grant itself is dead.
    ///
    /// `invalid_grant` is the only refusal that means the refresh token is spent and signing in
    /// again is the way back. Everything else a 400 can carry — `invalid_request`, a proxy's error
    /// page, a body that is not JSON at all — leaves a refresh token that is still good, so an
    /// unrecognised refusal reads as "not this", and the credential is kept.
    static func isInvalidGrant(_ body: String) -> Bool {
        struct Refusal: Decodable { let error: String? }
        return (try? JSONDecoder().decode(Refusal.self, from: Data(body.utf8)))?.error == "invalid_grant"
    }

    static func refresh(_ refreshToken: String) async throws -> TokenSet {
        try await token(form: [
            "grant_type": "refresh_token",
            "refresh_token": refreshToken,
            "client_id": Configuration.oauthClientID,
        ])
    }

    private static func token(form: [String: String]) async throws -> TokenSet {
        var request = URLRequest(url: Configuration.baseURL.appendingPathComponent("token"))
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        request.httpBody = Data(form.map { "\($0.key)=\($0.value.formURLEncoded)" }.joined(separator: "&").utf8)

        let (data, response) = try await URLSession.shared.data(for: request)
        let status = (response as? HTTPURLResponse)?.statusCode ?? 0
        guard (200..<300).contains(status) else {
            throw OAuthError.tokenEndpoint(status: status, body: String(decoding: data, as: UTF8.self))
        }

        let decoded = try JSONDecoder().decode(TokenResponse.self, from: data)
        // The refresh token rotates on every use server-side, so a response without one would leave
        // nothing to refresh with next time — better to fail here than to discover it in an hour.
        guard let refresh = decoded.refresh_token else { throw OAuthError.noRefreshToken }
        return TokenSet(
            accessToken: decoded.access_token,
            refreshToken: refresh,
            accessExpiresAt: Date().addingTimeInterval(decoded.expires_in ?? 3600)
        )
    }
}

enum OAuthError: LocalizedError {
    case notAuthenticated
    case tokenEndpoint(status: Int, body: String)
    case noRefreshToken
    case stateMismatch

    var errorDescription: String? {
        switch self {
        case .notAuthenticated:
            return "Not signed in to Helmsley. Open Helmsley Drive and sign in again."
        case .tokenEndpoint(let status, let body):
            return "The sign-in server refused the request (HTTP \(status)): \(body)"
        case .noRefreshToken:
            return "The sign-in server issued no refresh token."
        case .stateMismatch:
            return "The sign-in response did not match the request that started it."
        }
    }
}

/// Hands out an access token that is valid right now, refreshing when it is not.
///
/// An actor, but the actor is not what makes a refresh happen once: actors are reentrant, and every
/// `await` in here is a door another caller can walk through while the first is suspended on the
/// network. The extension produces exactly that traffic — a poll round and a handful of fetches all
/// asking for a token in the same breath — and the server rotates refresh tokens, so two calls
/// spending the same one would see the second revoke what the first had just been issued. What
/// serialises the rotation is `refreshing`: one task does the work, and everyone who arrives while
/// it runs awaits that task instead of starting another.
actor TokenProvider {

    static let shared = TokenProvider()

    private var cached: TokenSet?

    /// The refresh in flight, if any. See the note on the actor: this task, not actor isolation, is
    /// what guarantees the refresh token is spent once per rotation in this process.
    private var refreshing: Task<String, Error>?

    var isSignedIn: Bool { (cached ?? TokenStore.load()) != nil }

    func store(_ tokens: TokenSet) throws {
        try TokenStore.save(tokens)
        cached = tokens
    }

    func signOut() {
        TokenStore.clear()
        cached = nil
    }

    /// Drops the in-memory copy, so the next question is answered from the keychain.
    ///
    /// For the app's window when it comes back to the front: the keychain is the only store the
    /// extension writes, and a cached access token can go on answering for up to an hour after the
    /// extension has signed out — or rotated — behind this process's back. What the window shows
    /// should be the shared truth, not this process's memory of it.
    func resync() {
        cached = nil
    }

    func accessToken() async throws -> String {
        if let tokens = cached, tokens.isAccessTokenFresh { return tokens.accessToken }

        if let refreshing { return try await refreshing.value }
        let task = Task { try await refresh() }
        refreshing = task
        defer { refreshing = nil }
        return try await task.value
    }

    /// The slow path: nothing fresh in memory, so the keychain and then the server are asked. Runs
    /// as the one `refreshing` task; everything below stays on the actor.
    private func refresh() async throws -> String {
        // Re-read before refreshing: the other process may have rotated the pair since this one
        // last looked, and spending a refresh token it has already replaced would revoke the live
        // credential and sign the user out of both.
        guard let stored = TokenStore.load() else {
            // The extension reaching this is the difference between "the portal is empty" and "this
            // process cannot see the credential the app saved" — indistinguishable in Finder.
            Log.auth.error("no token in keychain (access group \(Configuration.keychainAccessGroup, privacy: .public))")
            throw OAuthError.notAuthenticated
        }
        if stored.isAccessTokenFresh {
            cached = stored
            return stored.accessToken
        }

        do {
            return try await rotate(from: stored)
        } catch OAuthError.tokenEndpoint(let status, let body)
            where (status == 400 || status == 401) && OAuth.isInvalidGrant(body) {
            // The server says the grant itself is dead — not that the request was malformed or that
            // a proxy hiccuped, which also arrive as 400s and must not cost the credential. Either
            // the other process won a simultaneous refresh — in which case the keychain now holds a
            // pair this one has never seen, and that pair is good — or there is nothing left but to
            // sign in again.
            if let latest = TokenStore.load(), latest.refreshToken != stored.refreshToken {
                if latest.isAccessTokenFresh {
                    cached = latest
                    return latest.accessToken
                }
                return try await rotate(from: latest)
            }
            signOut()
            throw OAuthError.notAuthenticated
        }
    }

    private func rotate(from tokens: TokenSet) async throws -> String {
        Log.auth.info("refreshing access token")
        let refreshed = try await OAuth.refresh(tokens.refreshToken)
        // In memory before the keychain: the server has already retired the pair this was minted
        // from, so if the save fails this copy is the only one anywhere — throwing it away over a
        // keychain hiccup would sign the user out while holding a perfectly good credential.
        cached = refreshed
        do {
            try TokenStore.save(refreshed)
        } catch {
            Log.auth.error("could not save the refreshed tokens — carrying on in memory: \(error.localizedDescription, privacy: .public)")
        }
        return refreshed.accessToken
    }
}

extension Data {
    /// RFC 7636 wants base64url with no padding, which `base64EncodedString()` is not.
    func base64URLEncodedString() -> String {
        base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }
}

extension String {
    var formURLEncoded: String {
        // `.urlQueryAllowed` leaves `+`, `&` and `=` intact, which in a form body would be read as
        // structure rather than as content.
        let unreserved = CharacterSet(charactersIn: "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~")
        return addingPercentEncoding(withAllowedCharacters: unreserved) ?? self
    }
}
